runtime/near-vm-logic/src/alt_bn128.rs

```diff
@@ -0,0 +1,249 @@
+    };
+
+    if res < B * n + discount {
+        return 0;
+    }
+    res -= B * n + discount;
+    res
+}
+
+#[derive(Copy, Clone)]
+struct WrapU256(pub U256);
+
+#[derive(Copy, Clone)]
+struct WrapFr(pub Fr);
+
+#[derive(Copy, Clone)]
+struct WrapFq(pub Fq);
+
+#[derive(Copy, Clone)]
+struct WrapFq2(pub Fq2);
+
+#[derive(Copy, Clone)]
+struct WrapG1(pub G1);
+
+#[derive(Copy, Clone)]
+struct WrapG2(pub G2);
+
+impl BorshSerialize for WrapU256 {
+    fn serialize<W: Write>(&self, writer: &mut W) -> Result<(), Error> {
+        self.0 .0.serialize(writer)
+    }
+}
+
+impl BorshDeserialize for WrapU256 {
+    fn deserialize(buf: &mut &[u8]) -> Result<Self, Error> {
+        let value = <[u128; 2]>::deserialize(buf)?;
+        Ok(WrapU256(U256(value)))
+    }
+}
+
+impl BorshSerialize for WrapFr {
+    fn serialize<W: Write>(&self, writer: &mut W) -> Result<(), Error> {
+        self.0.into_u256().serialize(writer)
+    }
+}
+
+impl BorshDeserialize for WrapFr {
+    fn deserialize(buf: &mut &[u8]) -> Result<Self, Error> {
+        let num = WrapU256::deserialize(buf)?.0;
+        Fr::new(num)
+            .ok_or_else(|| Error::new(ErrorKind::InvalidData, "integer is not less than modulus"))
+            .map(|r| WrapFr(r))
+    }
+}
+
+impl BorshSerialize for WrapFq {
+    fn serialize<W: Write>(&self, writer: &mut W) -> Result<(), Error> {
+        self.0.into_u256().serialize(writer)
+    }
+}
+
+impl BorshDeserialize for WrapFq {
+    fn deserialize(buf: &mut &[u8]) -> Result<Self, Error> {
+        let num = WrapU256::deserialize(buf)?.0;
+        Fq::from_u256(num)
+            .map_err(|_| Error::new(ErrorKind::InvalidData, "integer is not less than modulus"))
+            .map(|r| WrapFq(r))
+    }
+}
+
+impl BorshSerialize for WrapFq2 {
+    fn serialize<W: Write>(&self, writer: &mut W) -> Result<(), Error> {
+        self.0.real().serialize(writer)?;
+        self.0.imaginary().serialize(writer)
+    }
+}
+
+impl BorshDeserialize for WrapFq2 {
+    fn deserialize(buf: &mut &[u8]) -> Result<Self, Error> {
+        let re = WrapFq::deserialize(buf)?.0;
+        let im = WrapFq::deserialize(buf)?.0;
+
+        Ok(WrapFq2(Fq2::new(re, im)))
+    }
+}
+
+impl BorshSerialize for WrapG1 {
+    fn serialize<W: Write>(&self, writer: &mut W) -> Result<(), io::Error> {
+        match AffineG1::from_jacobian(self.0) {
+            Some(p) => {
+                p.x().serialize(writer)?;
+                p.y().serialize(writer)?;
+            }
+            None => {
+                Fq::zero().serialize(writer)?;
+                Fq::zero().serialize(writer)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl BorshDeserialize for WrapG1 {
+    fn deserialize(buf: &mut &[u8]) -> Result<Self, io::Error> {
+        let x = Fq::deserialize(buf)?;
+        let y = Fq::deserialize(buf)?;
+        if x.is_zero() && y.is_zero() {
+            Ok(WrapG1(G1::zero()))
+        } else {
+            AffineG1::new(x, y)
```

Comment:
It should not be abused, because gas is accounted before deserialization.
Gas cost of deserialization is close to gas metering for alt_bn128_g1_sum function.

Reply:
I see it now.